address #7586: make hintsSource print context for all msgs, add

timotheecour opened the pull request:
partially address https://github.com/nim-lang/Nim/issues/7586: make hintsSource print context for all msgs, add hintsSourceError to print context for errors (ie last entry if
compilation aborts on error); adjust these hints in verbosity; minor
code clarification


## NOTE:
also had to add `--hint[sourceError]:off` to tests/errmsgs/t5167_5.nim

Araq commented on compiler/msgs.nim:
```diff
@@ -134,7 +134,7 @@ type
     hintConditionAlwaysTrue, hintName, hintPattern,
     hintExecuting, hintLinking, hintDependency,
-    hintSource, hintPerformance, hintStackTrace, hintGCStats,
+    hintSource, hintSourceError, hintPerformance, hintStackTrace, hintGCStats,
```
Why do we need ``hinSourceError`` when ``hinSource`` already exists and was used for the same?

timotheecour replied:
after this PR:
hintSourceError only prints source code for errors
hintSource prints source code for all entries

hintSourceError only adds 2 lines in compilation messages (including caret `^`), because there can only be 1 error (IUUC) eg:
```
/Users/timothee/git_clone/nim/timn/tests/t25_hint_source.nim(6, 3) Error: undeclared identifier: 'a'
    a: int
    ^
```
so it's ok to have it on by default with default verbosity 1 (as done with this PR)

hintSource adds more lines to compilation messages, so I only enable it for `hints[source]`
```
/Users/timothee/git_clone/nim/timn/tests/t25_hint_source.nim(2, 39) Warning: expr is deprecated [Deprecated]
  macro foo1*(cond: untyped, msg = ""): expr = discard
                                        ^
/Users/timothee/git_clone/nim/timn/tests/t25_hint_source.nim(3, 39) Warning: expr is deprecated [Deprecated]
  macro foo2*(cond: untyped, msg = ""): expr = discard
                                        ^
/Users/timothee/git_clone/nim/timn/tests/t25_hint_source.nim(6, 3) Error: undeclared identifier: 'a'
    a: int
    ^
```

timotheecour commented:
@Araq some tests fail after I had made `hints[sourceError]` the default (ie via verbosity:1) because they expected a strict error message that doesn't contain line source contents.
I could've fixed these but maybe other libs would also need to be fixed, so I instead didn't change the default ; users can add `hints[source]` or `hints[sourceError]` if they want to

Araq commented:
> @Araq some tests fail after I had made hints[sourceError] the default (ie via verbosity:1) because they expected a strict error message that doesn't contain line source contents.

That's most a one-line fix in testament adapting the command it uses to invoke the compiler.

timotheecour commented:
@Araq PTAL.
I removed hintSourceError, made hints[source] default (ie enabled in verbosity:1), and updated `tests/testament/specs.nim` to make tests pass.
Note, the fix in tests/testament/specs.nim is a bit ugly, it'd be better to do via a nim.cfg that'd be read just for tests, but not sure how to do that (editing tests/testament/tester.nim.cfg didn't help)
Note, after this PR, the logs do increase in terms of number of lines (3 lines per compiler message instead of 1), so I really hope you can accept my other PR https://github.com/nim-lang/Nim/pull/7636 to fix that.

timotheecour commented:
so, what's missing to merge this PR? on/off/auto could be done in another PR

timotheecour commented:
PTAL, `hint[source]` now defaults at verbosity>2, I think it's cleaner that way rather than force it on the user (and fix other tools); if user has `hint[source]:on` in ~/.config/nim.cfg`, the change i made earlier to `tests/testament/specs.nim` should do the right thing to avoid breaking tests

Araq commented on compiler/msgs.nim:
```diff
@@ -1032,7 +1039,7 @@ proc liMessage(info: TLineInfo, msg: TMsgKind, arg: string,
       else:
         styledMsgWriteln(styleBright, x, resetStyle, color, title, resetStyle, s)
-      if msg in errMin..errMax and hintSource in gNotes:
+      if hintSource in gNotes:
```
This is now the only line that has any impact.

timotheecour replied:
not sure why you closed this; the point was to allow showing line contexts for every entry; as I wrote earlier (and also mentioned by @alehander42 here https://github.com/nim-lang/Nim/issues/7586#issuecomment-381096369), currently nim only shows line contexts for errors (which in practice would be the last entry), skipping all the contexts for warnings

Araq replied:
I applied this change manually.
